core/arch/arm/kernel/secure_partition.c

```diff
@@ -0,0 +1,776 @@
+
+	ns_buf_size = param->u[0].mem.size;
+	if (ns_buf_size > spc->ns_comm_buf_size)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	res = stmm_map_ns_buf(spc, param, &ns_buf_base);
+	if (res)
+		return res;
+
+	spc->regs.x[0] = FFA_MSG_SEND_DIRECT_REQ_64;
+	spc->regs.x[1] = (stmm_pta_id << 16) | stmm_id;
+	spc->regs.x[2] = FFA_PARAM_MBZ;
+	spc->regs.x[3] = spc->ns_comm_buf_addr;
+	spc->regs.x[4] = ns_buf_size;
+	spc->regs.x[5] = 0;
+	spc->regs.x[6] = 0;
+	spc->regs.x[7] = 0;
+
+	tee_ta_push_current_session(s);
+
+	memcpy((void *)(spc->ns_comm_buf_addr), (void *)ns_buf_base,
+	       ns_buf_size);
+
+	res = sec_part_enter_user_mode(spc);
+
+	/* Error handling TBD */
+	if (!res)
+		param->u[1].val.a = spc->regs.x[1];
+
+	memcpy((void *)ns_buf_base, (void *)(spc->ns_comm_buf_addr),
+	       ns_buf_size);
+
+	/*
+	 * Clear out the parameter mappings added with tee_mmu_map_param()
+	 * above.
+	 */
+	tee_mmu_clean_param(&spc->uctx);
+
+	tee_ta_pop_current_session();
+
+	return res;
+}
+
+static void stmm_enter_close_session(struct tee_ta_session *s __unused)
+{
+}
+
+static void sec_part_dump_state(struct tee_ta_ctx *ctx)
+{
+	user_mode_ctx_print_mappings(to_user_mode_ctx(ctx));
+}
+
+static uint32_t sec_part_get_instance_id(struct tee_ta_ctx *ctx)
+{
+	return to_sec_part_ctx(ctx)->uctx.vm_info.asid;
+}
+
+static void sec_part_ctx_destroy(struct tee_ta_ctx *ctx)
+{
+	struct sec_part_ctx *spc = to_sec_part_ctx(ctx);
+
+	tee_pager_rem_um_areas(&spc->uctx);
+	vm_info_final(&spc->uctx);
+	free(spc);
+}
+
+static int sp_svc_set_mem_attr(vaddr_t va, unsigned int nr_pages, uint32_t perm)
+{
+	TEE_Result res;
+	struct tee_ta_session *sess = NULL;
+	struct sec_part_ctx *spc = NULL;
+	uint32_t prot;
+
+	if (va == 0 || nr_pages == 0)
+		return SP_RET_INVALID_PARAM;
+
+	res = tee_ta_get_current_session(&sess);
+	if (res != TEE_SUCCESS)
+		return SP_RET_DENIED;
+
+	spc = to_sec_part_ctx(sess->ctx);
+
+	prot = 0;
+	if ((perm & SP_MEM_ATTR_ACCESS_MASK) == SP_MEM_ATTR_ACCESS_RW)
+		prot |= TEE_MATTR_URW | TEE_MATTR_PRW;
+	else if ((perm & SP_MEM_ATTR_ACCESS_MASK) == SP_MEM_ATTR_ACCESS_RO)
+		prot |= TEE_MATTR_UR | TEE_MATTR_PR;
+
+	if ((perm & SP_MEM_ATTR_EXEC_MASK) == SP_MEM_ATTR_EXEC) {
+		prot |= TEE_MATTR_UX;
+		prot &= ~(TEE_MATTR_UW | TEE_MATTR_PW);
+	}
+
+	res = vm_set_prot(&spc->uctx, va, nr_pages * SMALL_PAGE_SIZE, prot);
+	return res == TEE_SUCCESS ? SP_RET_SUCCESS : SP_RET_DENIED;
+}
+
+static bool return_helper(bool panic, uint32_t panic_code,
+			  struct thread_svc_regs *svc_regs)
+{
+	if (!panic) {
+		struct tee_ta_session *sess = NULL;
+		struct sec_part_ctx *spc = NULL;
+		size_t n = 0;
+
+		tee_ta_get_current_session(&sess);
+		spc = to_sec_part_ctx(sess->ctx);
+
+		spc->regs.x[0] = svc_regs->x0;
+		spc->regs.x[1] = svc_regs->x1;
+
+		/* Save the state to return to */
+		for (n = 19; n <= 29; n++)
+			spc->regs.x[n] = spc->regs.x[n];
+
+		/* In case user mode was AArch32 */
+		for (n = 5; n <= 14; n++)
+			spc->regs.x[n] = *(&svc_regs->x0 + n);
```

Comment:
Is this needed, given the dummy copy above?

Reply:
Both of those were part of the initial patch implementing secure partition loading from you. 
TBH i've never actually considered as why this was needed

Reply:
I had a closer look, this can be done better. Also with the new FFA changes we need to preserve more registers. I'll refactor it for the actual patchset.